### PR TITLE
Enable `dense` keyword in `ParametrizedEvolution`

### PR DESCRIPTION
**Pull request by Qottmann**
Hot fix for https://github.com/PennyLaneAI/pennylane/issues/4078 (see also https://github.com/google/jax/issues/15844)

Enables a keyword `dense` to be passed to `ParametrizedEvolution` that enforces dense evolution, which currently automatically is enabled for more than 2 wires

Edit2: will have to wait for jax 0.4.9 support to enable BCSR, so this is a hotfix for the moment (see discussion)

### CODE REVIEW

**Comment by Qottmann** [comment]
Implemented the fix from https://github.com/google/jax/issues/15844 and https://github.com/PennyLaneAI/pennylane/issues/4078

In case that solves it (and doesnt cause any unexpected behavior / failing tests), would we still want the possibility to enforce dense matrices?

File: pennylane/pulse/parametrized_evolution.py [hunk 1]
```diff
@@ -393,6 +396,7 @@ def __init__(
         self.hyperparameters["return_intermediate"] = return_intermediate
         self.hyperparameters["complementary"] = complementary
         self._check_time_batching()
+        self.dense = dense or len(self.wires) < 3
```
Comment (PhilipVinc) [inline pennylane/pulse/parametrized_evolution.py:399]: I don't personally like to rely on the implicit False-ness of `None` so I'd suggest this more explicit usage, but I don't know what's the standard style in PL..

```suggestion
        if dense is None:
            dense = len(self.wires) < 3
        self.dense = dense
```

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
Can we also have a test checking that specifying `dense=True` works?

**Comment by PhilipVinc** [comment]
> cause any unexpected behavior / failing tests), would we still want the possibility to enforce dense matrices?

I'd say that's useful. Can be deprecated in the future if we don't want this to be an option anymore.

**Comment by Qottmann** [comment]
This is blocked because we are stuck at `0.4.3` jax version. The main problem has been fixed on the jax side in https://github.com/google/jax/issues/14833 but we would also need to drop python 3.8 and that may take a bit longer 🥲 CC @rmoyard

**Comment by mlxd** [comment]
Just a fly-by comment: once this PR is ready to go can we enable the GPU CI tests by adding the "gpu" label to this PR?
To avoid overheads, I'd consider this only when the work is ready to merge.

**Comment by rmoyard** [comment]
The GPU tests for the moment only have Torch.

**Comment by mlxd** [comment]
Yep, as @rmoyard said only torch is supported now. We should ensure the runner installs the JAX CUDA version for this to be supported.

File: pennylane/pulse/parametrized_evolution.py (hunk 1 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py:399]: This is quite standard in PL

File: pennylane/pulse/parametrized_evolution.py [hunk 2]
```diff
@@ -93,6 +93,8 @@ class ParametrizedEvolution(Operation):
             where :math:`t_i` are the additional times provided in ``t``.
             If ``True``, the *remaining* time evolution to :math:`t_f` is computed instead, returning
             :math:`\{U(t_0, t_f), U(t_1, t_f),\dots, U(t_{f-1}, t_f), U(t_f, t_f)\}`.
+        dense (bool): Whether the evolution should use dense matrices. Per default, this is decided by
+            the number of wires, i.e. ``dense = len(wires) < 3``.
```
Comment (dwierichs) [inline pennylane/pulse/parametrized_evolution.py:97]: I guess strictly speaking `wires` is not "defined" at this level? maybe:
```suggestion
            the number of wires, i.e. ``dense = len(H.wires) < 3``.
```

File: pennylane/pulse/parametrized_evolution.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/pulse/parametrized_evolution.py:399]: Wait, but this does not work if we actually want `dense=False` to be a valid value of the variable, right? :sweat_smile: 
```suggestion
        self.dense = len(self.wires) < 3 if dense is None else dense
```

File: tests/pulse/test_parametrized_evolution.py [hunk 3]
```diff
@@ -89,6 +89,18 @@ def test_init(self):
...
+    def test_set_dense(self):
+        """Test that flag dense is set correctly"""
+        ops = [qml.PauliX(0), qml.PauliY(1), qml.PauliZ(2)]
+        coeffs = [1, 2, 3]
+        H = ParametrizedHamiltonian(coeffs, ops)
+        ev = ParametrizedEvolution(H=H, params=[1, 2], t=2)
+        assert ev.dense is False
+
+        ev2 = ParametrizedEvolution(H=H, params=[1, 2], t=2, dense=True)
+        assert ev2.dense is True
```
Comment (dwierichs) [inline tests/pulse/test_parametrized_evolution.py:103]: Adding tests for less than three wires and `dense=None` as well as for `dense=False`.
```suggestion
        assert ev2.dense is True
        ops = [qml.PauliX(0), qml.PauliY(1)]
        coeffs = [1, 2]
        H = ParametrizedHamiltonian(coeffs, ops)
        ev = ParametrizedEvolution(H=H, params=[1, 2], t=2)
        assert ev.dense is True

        ev2 = ParametrizedEvolution(H=H, params=[1, 2], t=2, dense=False)
        assert ev2.dense is False
```
btw: Where are those `params` going? :o

**Review by dwierichs — COMMENTED** [review]
Thanks for the quick fix @Qottmann  :muscle: 
Three small comments :)

File: pennylane/pulse/parametrized_evolution.py (hunk 1 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py:399]: please ignore my ignorance, you are both right 🙈

File: pennylane/pulse/parametrized_evolution.py (hunk 2 above)
Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py:97]: It is defined in the `super().__init__`

Comment (Qottmann) [reply pennylane/pulse/parametrized_evolution.py:97]: But yes what you suggest is clearer 👍

File: tests/pulse/test_parametrized_evolution.py (hunk 3 above)
Comment (Qottmann) [reply tests/pulse/test_parametrized_evolution.py:103]: added a test with explicit dense=False, the case of 2 wires is tested in the init, do you want an explicit extra test with 2 wires?

Comment (dwierichs) [reply tests/pulse/test_parametrized_evolution.py:103]: Ah no, that's fine :)

**Review by dwierichs — APPROVED** [review]
Thanks @Qottmann !

**Comment by Qottmann** [comment]
Merging this and waiting with the change from csr to batched csr format when we have jax `0.4.9` supported in PL
